Sandbox/src/Game/Npc/NpcRepository.cpp

```diff
@@ -0,0 +1,197 @@
+#include "Game/Npc/NpcRepository.h"
+
+NpcRepository& NpcRepository::Instance()
+{
+	static NpcRepository instance;
+	return instance;
+}
+
+NpcRepository::NpcRepository()
+{
+	Init();
+}
+
+NpcRepository::~NpcRepository()
+{
+	for (auto npc : mNpcDefinitions)
+		delete npc.second;
+
+	for (auto npc : mNpcCombatDefinitions)
+		delete npc.second;
+
+	for (auto npc : mNpcDropDefinitions)
+		delete npc.second;
+
+	for (auto npc : mNpcShopDefinitions)
+		delete npc.second;
+}
+
+void NpcRepository::Init()
+{
+	LoadNpcDefinitions();
+	LoadNpcCombatDefinitions();
+	LoadNpcDropDefinitions();
+	LoadNpcShopDefinitions();
+}
+
+void NpcRepository::LoadNpcDefinitions()
+{
+	std::string fullPath = Square::AssetManager::Instance().GetBasePath();
+	fullPath.append("Assets/Definitions/");
+
+	std::ifstream infile(fullPath + "NpcDefinition.dat");
+	if (!infile) std::cerr << "NpcDefinition.dat could not be opened!" << std::endl;
+	else std::cout << "Loading npc definitions..." << std::endl;
+
+	int index;
+	std::string name;
+	std::string spritePath;
+	std::string actions;
+
+	while (infile >> index >> std::quoted(name) >>  spritePath >> std::quoted(actions))
+	{
+		mNpcDefinitions[index] = new NpcDefinition(index, name, spritePath, DelimitedStringToVector(actions));
+	}
+
+	infile.close();
+	std::cout << "Npc definitions loaded: " << mNpcDefinitions.size() << std::endl;
+}
+
+void NpcRepository::LoadNpcCombatDefinitions()
+{
+	std::string fullPath = Square::AssetManager::Instance().GetBasePath();
+	fullPath.append("Assets/Definitions/");
+
+	std::ifstream infile(fullPath + "NpcCombatDefinition.dat");
+	if (!infile) std::cerr << "NpcCombatDefinition.dat could not be opened!" << std::endl;
+	else std::cout << "Loading npc combat definitions..." << std::endl;
+
+	std::string line;
+
+	while (std::getline(infile, line))
+	{
+		std::istringstream iss(line);
+		int index;
+		float respawnTime;
+		bool aggressive;
+		bool poisonous;
+		int level;
+		int hitpoints;
+		int maxHit;
+		float attackSpeed;
+		int combatStyle;
+		int weakness;
+		std::array<int, 5> stats;
```

Comment:
maybe put `5` into a constant somewhere since it's being used in a lot of places

Reply:
That's fair lol

Reply:
Once I work on the combat code, I'll probably end up refactoring this class and the player. If I can get them to inherit from the same class I can avoid writing duplicate code for when they attack each other and just have the underlying logic be the same